packages/microcosm/src/observable.js

```diff
@@ -181,105 +159,136 @@ export class Observable {
       for (var i = 0; i < arguments.length; ++i) {
         observer.next(arguments[i])
       }
-
-      observer.complete()
-    })
-  }
-
-  static wrap(source: *): Observable {
-    if (source && typeof source === 'object') {
-      let builder = getObservable(source)
-
-      if (builder) {
-        return builder.call(source)
-      }
-
-      if (typeof source.then === 'function') {
-        return Observable.fromPromise(source)
-      }
-    }
-
-    return new Observable(observer => {
-      observer.next(source)
       observer.complete()
     })
   }
-
-  static fromPromise(promise: Promise<*>): Observable {
-    return new Observable(observer => {
-      promise
-        .then(observer.next)
-        .then(observer.complete)
-        .catch(observer.error)
-    })
-  }
 }
 
-export function getObservable(obj: any) {
-  return obj && obj[observable]
+function disposeSubscription(subscription: Subscription) {
+  closeSubscription(subscription)
+  cleanupSubscription(subscription)
 }
 
-function cleanupSubscription(subscription) {
+function cleanupSubscription(subscription: Subscription): void {
   let cleanup = subscription._cleanup
 
-  subscription._origin._subscriptions.delete(subscription)
-
-  if (cleanup) {
-    // Drop the reference to the cleanup function so that we won't call it
-    // more than once
-    subscription._cleanup = undefined
+  // Drop the reference to the cleanup function so that we won't call it
+  // more than once
+  subscription._cleanup = noop
 
-    // Call the cleanup function
+  if (typeof cleanup === 'function') {
     cleanup()
+  } else if (cleanup instanceof Subscription) {
+    cleanup.unsubscribe()
   }
 }
 
-function handleNext(subscription: Subscription, value?: *) {
-  if (subscription._observer) {
-    subscription._observer.next(value)
-  }
-}
+const ClosedObserver = new Observer()
 
-function handleComplete(subscription: Subscription): void {
-  let observer = subscription._observer
+function closeSubscription(subscription: Subscription): void {
+  subscription._observer = ClosedObserver
+  subscription._queue.length = 0
+  subscription._state = CLOSED
+}
 
-  subscription.close()
+function flushSubscription(subscription: Subscription): void {
+  let queue = subscription._queue
 
-  observer.complete()
+  subscription._queue = []
+  subscription._state = WORKING
 
-  cleanupSubscription(subscription)
-}
+  for (let i = 0; i < queue.length; ++i) {
```

Comment:
would `++i` increment `i` to `1` before the block is called?

Reply:
Nope! The third statement is the post-condition of the loop. This is just confusing. I'll replace it with `i++`, which I should have had to begin with.